### DISCUSSION TITLE
Bash Filter with .not-Logic possible?

### DISCUSSION THREAD

**Discussion by livingmydesign** [supabase/supabase#11843] [Questions] [answered]
**Dear Community,**

as I could not find any examples that could relate to my case and even AI was not able to solve it, I hope to find some support in this community on this simple but still for me very difficult bash command.

I need a bash command for a supabase api GET, that results in a **list of rows which DO-NOT contain** the string variable '**uid**'  in the json-array columns "**Yes_voters**" and "**No_voters**". 

![Screenshot 2023-01-22 at 17 35 31](https://user-images.githubusercontent.com/93006467/213927612-b384f530-8af4-4087-989e-494a45f32b0a.png)

in the supabase docs I was able to find further bash docs for filters besides of the 'eq.'
![Screenshot 2023-01-22 at 17 37 28](https://user-images.githubusercontent.com/93006467/213928166-6b24d121-9c27-4ca5-9b10-c56190d120cb.png)

the other operators listed here, also seem not to work for my specific case:
https://postgrest.org/en/stable/api.html#operators

After trying to get a solution for the whole day, I dearly hope that someone of you can give me the essential hint.

Thank you for your time!

**Comment by GaryAustin1** [comment, marked as answer]
I'm not sure about contains (cs) and json array, but the logic seems to be there for your logic...
`?not.and=(a.gte.0,a.lte.100)`   Where you would use `col.cs.{123,456}`
https://postgrest.org/en/stable/api.html#logical-operators
There is also .or and you can nest them.

**Reply by GaryAustin1** [reply]
Try not to create new answers to comment on an answer...

In your url below,  I don't see parenthesis like `=(Yes_voters.cs.{221968}) `

**Reply by GaryAustin1** [reply]
Also see the discussion here about jsonb versus jsonb[].   Not sure why you need a jsonb column if you have array of text (use text[]), or if you use jsonb you may not want an array type.

https://github.com/supabase/supabase/discussions/4784

**Reply by livingmydesign** [reply]
Thank you Gary for your support!
I am very new to the bash coding, therefore my questions might sound obvious to answer for others.
Just trying to get a bash that returns only the middle col where the '221968' is not present.

![Screenshot 2023-01-23 at 00 30 34](https://user-images.githubusercontent.com/93006467/213946352-53155986-169f-40f5-9da6-afa8ee247169.png)

when I use jsonb, at least I get results. but still it gives me the opposite result. 
and when I just remove the "not" it gives me no result. it seems I got code-blind and can't see the obvious answer, that is probably just in front I suppose.
![Screenshot 2023-01-23 at 00 31 21](https://user-images.githubusercontent.com/93006467/213946395-b35801da-917d-4fb0-9550-4b02627e3fdc.png)

**Reply by livingmydesign** [reply]
I kept on trying with all different scenarios and datatypes meanwhile and tend to come to conclusion that Supabase does not fully support postgrest cs. and cd. on arrays..

@GaryAustin1

**Reply by GaryAustin1** [reply]
If you are using URL's directly then you are going directly to PostgREST so anything PostgREST can do Supabase can do.  The only time Supabase is involved is if you use one of their clients (like supabase-js).

edit: If you meant Postgres (no t) then I'm not sure as I've not used the SQL syntax for it.

I have used cs before on text arrays (text[]) for tags and it works as expected.

**Reply by livingmydesign** [reply]
yes, it should be as a postgresT bash command

let me show you this example, where I tried to get the cs. operator right on text[] arrays:
![bash_test](https://user-images.githubusercontent.com/93006467/214926760-c1db5065-a3ab-4364-8213-e10c10c0ace2.png)

**Reply by livingmydesign** [reply]
Ahh... it seems to have an issue with the NULL in the Arrays.
I just tested it with [""] for empty arrays and it seems to work now.

**Reply by GaryAustin1** [reply]
Are you asking why .not is leaving out NULL rows?  If so, those rows don't exist at all when using a Postgres Where statement to evaluate values (which is what PostgREST translates to).

You probably need to add an .or filter with .is.NULL to pick up those rows.

https://stackoverflow.com/questions/44217155/using-null-in-where-clause

**Reply by livingmydesign** [reply]
Well I was not able to make a .NULL statement work for a Postgrest operation but I decided to change the supabase defaultvalue to {} which then seems to work somehow. 

thanks for your hints along the way @GaryAustin1 !

**Reply by GaryAustin1** [reply]
{} empty array is a "value" it can be filtered and compared against.  Null cannot except using IS.  It is how postgres the database works.  Filtering on a column will simply ignore rows with null in that column.   
Example row1 has null, row2 has 'a', row3 has ''
filtering on col != 'a'  will return row3 but not row1.
